### DISCUSSION TITLE
Introducing new variables in custom loss function

### DISCUSSION THREAD

**Discussion by quasars100** [MilesCranmer/PySR#385] [Q&A] [answered]
Hi Dr. Cranmer (or others),

For the PySR custom loss function (loss = "loss(x,y, w[optional]) = ") we specify in the model, right now it seems "x" is the data input while "y" is the analytic value found by PySR, and w are any weights the data has (which is optional), though please correct me if I misunderstood the variables (and this could potentially solve my problem).  However, if I'm using PySR to model something very specific (i.e. Poisson likelihood) and have a custom loss function that also depends on other variables ( data x = x(r), where I also want to include r in the loss function), is there a way to do this?  https://astroautomata.com/PySR/options/#loss has some nice examples of loss functions, but not introducing any other variables than the data "x" and fit value "y".
Thank you.

**Comment by MilesCranmer** [comment]
Hi @quasars100,

The loss argument is of the form `loss(prediction, target)`. The evaluation used to create the predictions happens automatically using your passed data (a matrix of shape (examples,features)) and the current expression. Let me know if I misinterpreted your question though!

cheers,
Miles

**Comment by MilesCranmer** [comment]
Oh wait I think I misunderstood. Do you need weights *and* the radius r? If you only need the radius in the loss function, you can just treat it like `loss(prediction, target, r)=…`, and pass `weights=r`. But if you need both, will be trickier. Let me know.

**Reply by quasars100** [reply]
Hi Miles,
Thank you so much for your reply!  I only need the radius, where target = target(r) --> the target data points I am trying to fit to is a function of the radius r.  Are weights here just any additional free parameters we can use in the loss function besides the target (data points) and prediction (analytic function found by PySR)?  If that's the case, then the solution you proposed would work perfectly (sorry I thought weights had some intrinsic definition similar to statistical weights).

**Reply by MilesCranmer** [reply, marked as answer]
>  Thank you so much for your reply! I only need the radius, where target = target(r) --> the target data points I am trying to fit to is a function of the radius r. Are weights here just any additional free parameters we can use in the loss function besides the target (data points) and prediction (analytic function found by PySR)? If that's the case, then the solution you proposed would work perfectly (sorry I thought weights had some intrinsic definition similar to statistical weights).

Exactly! 

They are called "weights" because the standard loss function treats them as regular statistical weightings. But if you are writing a custom loss function anyways, you can use them however you like.
